go/pipeline/batch.go

```diff
@@ -0,0 +1,215 @@
+type BatchOptions struct {
+	Timeout       *uint32
+	Route         *config.Route
+	RetryStrategy *ClusterBatchRetryStrategy
+}
+
+func (sbo StandaloneBatchOptions) Convert() BatchOptions {
+	return BatchOptions{Timeout: sbo.Timeout}
+}
+
+func (cbo ClusterBatchOptions) Convert() BatchOptions {
+	return BatchOptions{Timeout: cbo.Timeout, Route: cbo.Route, RetryStrategy: cbo.RetryStrategy}
+}
+
+// ====================
+
+// TODO make private if possible
+type Batch struct {
+	Commands []Cmd
+	IsAtomic bool
+	// TODO make private
+	Converters []func(any) any
+}
+
+type BaseBatch[T StandaloneBatch | ClusterBatch] struct {
+	Batch
+	self *T
+}
+
+type StandaloneBatch struct {
+	BaseBatch[StandaloneBatch]
+}
+
+type ClusterBatch struct {
+	BaseBatch[ClusterBatch]
+}
+
+// ====================
+
+func (b Batch) Convert(response []any) ([]any, error) {
+	if len(response) != len(b.Converters) {
+		return nil, &errors.RequestError{Msg: "Converters misaligned"}
+	}
+	for i, res := range response {
+		response[i] = b.Converters[i](res)
+	}
+	return response, nil
+}
+
+// ====================
+
+func NewStandaloneBatch(isAtomic bool) *StandaloneBatch {
+	b := StandaloneBatch{BaseBatch: BaseBatch[StandaloneBatch]{Batch: Batch{IsAtomic: isAtomic}}}
+	b.self = &b
+	return &b
+}
+
+func NewClusterBatch(isAtomic bool) *ClusterBatch {
+	b := ClusterBatch{BaseBatch: BaseBatch[ClusterBatch]{Batch: Batch{IsAtomic: isAtomic}}}
+	b.self = &b
+	return &b
+}
+
+// Add a cmd to batch without response type checking nor conversion
+func (b *BaseBatch[T]) addCmd(request C.RequestType, args []string) *T {
+	b.Commands = append(b.Commands, Cmd{RequestType: request, Args: args})
+	b.Converters = append(b.Converters, func(res any) any { return res })
+	return b.self
+}
+
+// Add a cmd to batch with type checker but without response type conversion
+func (b *BaseBatch[T]) addCmdAndTypeChecker(
+	request C.RequestType,
+	args []string,
+	expectedType reflect.Kind,
+	isNilable bool,
+) *T {
+	return b.addCmdAndConverter(request, args, expectedType, isNilable, func(res any) any { return res })
+}
+
+// Add a cmd to batch with type checker and with response type conversion
+func (b *BaseBatch[T]) addCmdAndConverter(
+	request C.RequestType,
+	args []string,
+	expectedType reflect.Kind,
+	isNilable bool,
+	converter func(res any) any,
+) *T {
+	converterAndTypeChecker := func(res any) any {
+		if res == nil {
+			if isNilable {
+				return nil
+			}
+			return &errors.RequestError{
+				Msg: fmt.Sprintf("Unexpected return type from Glide: got nil, expected %v", expectedType),
+			}
+		}
+		if reflect.TypeOf(res).Kind() == expectedType {
+			return converter(res)
+		}
+		return &errors.RequestError{
+			Msg: fmt.Sprintf("Unexpected return type from Glide: got %v, expected %v", reflect.TypeOf(res), expectedType),
+		}
+	}
+	b.Commands = append(b.Commands, Cmd{RequestType: request, Args: args})
+	b.Converters = append(b.Converters, converterAndTypeChecker)
```

Comment:
Shouldn't converter just be a field (perhaps optional, if there is no conversion) on `Cmd` so that you don't need to slices that need to be in alignment?

Reply:
Fixed in b60a20724, thanks